base/precompilation.jl

```diff
@@ -564,7 +591,36 @@ function _precompilepkgs(pkgs::Vector{String},
         end
     end
     if !isempty(circular_deps)
-        @warn """Circular dependency detected. Precompilation will be skipped for:\n  $(join(string.(circular_deps), "\n  "))"""
+        outer_deps = copy(circular_deps)
+        cycles_names = ""
+        for cycle in cycles
+            filter!(!in(cycle), outer_deps)
+            cycle_str = ""
+            for (i, pkg) in enumerate(cycle)
+                j = max(0, i - 1)
+                if i == 1
+                    line = " ┌ "
+                elseif i < length(cycle)
+                    line = " │ " * " " ^j
+                else
+                    line = " └" * "─" ^j * " "
+                end
+                line = color_string(line, :light_black) * full_name(exts, pkg) * "\n"
+                cycle_str *= line
+            end
+            cycles_names *= cycle_str
+        end
+        plural1 = length(cycles) > 1 ? "these cycles" : "this cycle"
+        plural2 = length(outer_deps) > 1 ? "have" : "has"
+        plural3 = length(cycles) > 1 ? "cycles" : "cycle"
+        msg = """Circular dependency detected.
+        Precompilation will be skipped for dependencies all in $plural1:
+        $cycles_names"""
+        if !isempty(outer_deps)
+            msg *= "Precompilation will also be skipped for the following, which $plural2 depdenencies in the $plural3:\n"
```

Comment:
```suggestion
            msg *= "Precompilation will also be skipped for the following, which depend on the above $plural3:\n"
```

I find this wording clearer, but the original is fine too - just make sure to fix the spelling of `dependencies`